<!doctype html><html lang=en><head><meta charset=utf-8><meta http-equiv=x-ua-compatible content="IE=edge"><meta name=viewport content="width=device-width,initial-scale=1,shrink-to-fit=no"><title>exception - 小豪小豪</title><meta name=description content="exception"><meta name=author content="小豪"><link href=https://haomega.github.io/myblog/css/stylesheet.min.c5f3513f5b80c2549e72d959bdae43941f36c204c7504248d1e05f5fa480f14a.css integrity="sha256-xfNRP1uAwlSectlZva5DlB82wgTHUEJI0eBfX6SA8Uo=" rel="preload stylesheet" as=style><link rel=apple-touch-icon href=https://haomega.github.io/myblog/apple-touch-icon.png><link rel=icon href=https://haomega.github.io/myblog/favicon.ico><meta name=generator content="Hugo 0.74.3"><link rel=alternate type=application/rss+xml href=https://haomega.github.io/myblog/tags/exception/index.xml title=小豪小豪></head><body class=list id=top><header class=header><nav class=nav><p class=logo><a href=https://haomega.github.io/myblog/>小豪小豪</a></p><ul class=menu id=menu onscroll=menu_on_scroll()><li><a href=https://haomega.github.io/myblog/archives/>Archives</a></li><li><a href=https://haomega.github.io/myblog/tags/>Tags</a></li></ul></nav></header><main class=main><header class=page-header><h1>exception</h1></header><article class=post-entry><header class=entry-header><h2>Exception 体系概要</h2></header><section class=entry-content><p>概要 程序并不是每次都能按照理想情况运行，总是有着不可预测的异常情况发生。
作用 正确处理异常可以使程序更健壮； 异常给方法提供了除了return之外的额外的返回途径； Java中的异常结构 Java中的异常大致分为两类：Exception和Error，他们都继承自Throwable，意思是可抛出的异常;
Exception:(checked) 往往可以通过正确处理来避免；例如NullPointorException * runtimeException：运行时异常（unchecked） Error：（unchecked） 程序无能为力的异常；例如NoSuchClassError、OutOfmemoryError
异常的处理 有两种方式来处理异常
抛出 捕获 当你的的程序在某些情况不能正确执行：你可以抛出一个异常来警告调用者，或者捕获异常并处理。 当你的程序调用一个抛出异常的接口时： * 继续抛出这个异常 * 捕获并处理它（切记不可捕获了异常时不做任何处理） * 捕获它，doSomething并继续抛出此异常或另一个异常
异常处理原则 能用if/else或者其他能处理的，不要用异常 尽早抛出异常 带出重要信息 一旦捕获异常除非特别的，必须进行处理或者重新抛出。 ···· 其他 不捕获，一直抛出会发生什么？ 异常会击穿方法，直到被捕获，如果一直到最外层都没有捕获，则会杀掉当前线程。
catch/finally里的return finally一定会被执行，所以最终return的是finally块里的，另：不要再catch里return，这是不好的做法。
日志里的caused by是什么？ 出现caused by是异常被捕获后，重新由新的异常包裹（throw new XXXException(“some things”, e)）并抛出。...</p></section><footer class=entry-footer><time>May 28, 2020</time>
.
1min read
. 小豪</footer><a class=entry-link aria-label="post link to Exception 体系概要" href=https://haomega.github.io/myblog/posts/exception-%E6%A6%82%E8%BF%B0/></a></article></main><footer class=footer><span>&copy; 2020 <a href=https://haomega.github.io/myblog/>小豪小豪</a></span>
<span>&#183;</span>
<span>Powered by <a href=https://gohugo.io/ rel=noopener target=_blank>Hugo️️</a>️</span>
<span>&#183;</span>
<span>Theme️ <a href=https://github.com/adityatelange/hugo-PaperMod rel=noopener target=_blank>PaperMod</a></span></footer><a href=#top aria-label="go to top"><button class=top-link id=top-link type=button><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6"><path d="M12 6H0l6-6z"/></svg></button></a>
<script src=https://haomega.github.io/myblog/js/highlight.min.min.be617795b40f52d6ca2248c7fbf433c49211c7acfe8f221ac54f6b0007c66ea1.js integrity="sha256-vmF3lbQPUtbKIkjH+/QzxJIRx6z+jyIaxU9rAAfGbqE="></script><script>window.onload=function(){if(localStorage.getItem("menu-scroll-position")){document.getElementById('menu').scrollLeft=localStorage.getItem("menu-scroll-position");}}
hljs.initHighlightingOnLoad();document.querySelectorAll('a[href^="#"]').forEach(anchor=>{anchor.addEventListener("click",function(e){e.preventDefault();document.querySelector(this.getAttribute("href")).scrollIntoView({behavior:"smooth"});});});var mybutton=document.getElementById("top-link");window.onscroll=function(){if(document.body.scrollTop>800||document.documentElement.scrollTop>800){mybutton.style.visibility="visible";mybutton.style.opacity="1";}else{mybutton.style.visibility="hidden";mybutton.style.opacity="0";}};function menu_on_scroll(){localStorage.setItem("menu-scroll-position",document.getElementById('menu').scrollLeft);}</script></body></html>